File: backend/functions/src/recovery/recoveryService.test.ts
import {
  sleepComponent,
  trainingLoadComponent,
  weeklyPulseComponent,
  gutFeelingComponent,
  calculateScore,
  calculateComponents,
  recommendationForScore,
  RecoveryScoreComponents,
} from './recoveryService';

// Fixed "now" for deterministic date math
const NOW = new Date('2024-06-15T12:00:00Z');
const daysAgo = (n: number) =>
  new Date(NOW.getTime() - n * 24 * 60 * 60 * 1000);

describe('sleepComponent', () => {
  it('returns 50 when no logs', () => {
    expect(sleepComponent([])).toBe(50);
  });

  it('maps quality 1 to 0', () => {
    expect(sleepComponent([{ quality: 1 }])).toBe(0);
  });

  it('maps quality 5 to 100', () => {
    expect(sleepComponent([{ quality: 5 }])).toBe(100);
  });

  it('maps quality 3 to 50', () => {
    expect(sleepComponent([{ quality: 3 }])).toBe(50);
  });

  it('averages multiple logs', () => {
    expect(sleepComponent([{ quality: 1 }, { quality: 5 }])).toBe(50);
  });
});

describe('trainingLoadComponent', () => {
  it('returns 75 when no sessions', () => {
    expect(trainingLoadComponent([], NOW)).toBe(75);
  });

  it('returns 75 when all sessions are outside the 7-day window', () => {
    const sessions = [{ status: 'completed', date: daysAgo(10) }];
    expect(trainingLoadComponent(sessions, NOW)).toBe(75);
  });

  it('returns 50 when only recent load and no prior baseline', () => {
    const sessions = [{ status: 'completed', date: daysAgo(1) }];
    expect(trainingLoadComponent(sessions, NOW)).toBe(50);
  });

  it('gives higher score when recent load is lower than baseline', () => {
    // Heavy previous 4 days, nothing recent
    const heavy = [
      { status: 'completed', date: daysAgo(4) },
      { status: 'completed', date: daysAgo(5) },
      { status: 'completed', date: daysAgo(6) },
      { status: 'completed', date: daysAgo(7) },
    ];
    // Light last 3 days, same prior history
    const light = [
      { status: 'completed', date: daysAgo(1) },
      ...heavy,
    ];
    const highRecovery = trainingLoadComponent(heavy, NOW);
    const lowerRecovery = trainingLoadComponent(light, NOW);
    expect(highRecovery).toBeGreaterThan(lowerRecovery);
  });

  it('ignores planned sessions', () => {
    const sessions = [{ status: 'planned', date: daysAgo(1) }];
    expect(trainingLoadComponent(sessions, NOW)).toBe(75);
  });
});

describe('weeklyPulseComponent', () => {
  it('returns 50 when no pulses', () => {
    expect(weeklyPulseComponent([])).toBe(50);
  });

  it('maps all 5s to 100', () => {
    expect(
      weeklyPulseComponent([
        { energyScore: 5, sorenessScore: 5, motivationScore: 5, sleepQualityScore: 5 },
      ]),
    ).toBe(100);
  });

  it('maps all 1s to 0', () => {
    expect(
      weeklyPulseComponent([
        { energyScore: 1, sorenessScore: 1, motivationScore: 1, sleepQualityScore: 1 },
      ]),
    ).toBe(0);
  });

  it('maps all 3s to 50', () => {
    expect(
      weeklyPulseComponent([
        { energyScore: 3, sorenessScore: 3, motivationScore: 3, sleepQualityScore: 3 },
      ]),
    ).toBe(50);
  });
});

describe('gutFeelingComponent', () => {
  it('returns 50 when no meals', () => {
    expect(gutFeelingComponent([])).toBe(50);
  });

  it('maps stomachFeeling 1 to 0', () => {
    expect(gutFeelingComponent([{ stomachFeeling: 1 }])).toBe(0);
  });

  it('maps stomachFeeling 5 to 100', () => {
    expect(gutFeelingComponent([{ stomachFeeling: 5 }])).toBe(100);
  });
});

describe('calculateScore', () => {
  it('weights are 30/40/20/10', () => {
    const components: RecoveryScoreComponents = {
      sleep: 100,
      trainingLoad: 0,
      weeklyPulse: 0,
      gutFeeling: 0,
    };
    expect(calculateScore(components)).toBeCloseTo(30);
  });

  it('all 100 produces 100', () => {
    const c: RecoveryScoreComponents = {
      sleep: 100,
      trainingLoad: 100,
      weeklyPulse: 100,
      gutFeeling: 100,
    };
    expect(calculateScore(c)).toBe(100);
  });

  it('all 0 produces 0', () => {
    const c: RecoveryScoreComponents = {
      sleep: 0,
      trainingLoad: 0,
      weeklyPulse: 0,
      gutFeeling: 0,
    };
    expect(calculateScore(c)).toBe(0);
  });

  it('clamps to 100 on overflow', () => {
    const c: RecoveryScoreComponents = {
      sleep: 200,
      trainingLoad: 200,
      weeklyPulse: 200,
      gutFeeling: 200,
    };
    expect(calculateScore(c)).toBe(100);
  });
});

describe('recommendationForScore', () => {
  it('75 returns green recommendation', () => {
    expect(recommendationForScore(75)).toContain('well-recovered');
  });

  it('74 returns yellow recommendation', () => {
    expect(recommendationForScore(74)).toContain('Mild fatigue');
  });

  it('50 returns yellow recommendation', () => {
    expect(recommendationForScore(50)).toContain('Mild fatigue');
  });

  it('49 returns red recommendation', () => {
    expect(recommendationForScore(49)).toContain('High fatigue');
  });

  it('0 returns red recommendation', () => {
    expect(recommendationForScore(0)).toContain('High fatigue');
  });
});

describe('calculateComponents', () => {
  it('returns components struct with all four fields', () => {
    const c = calculateComponents([], [], [], []);
    expect(c).toHaveProperty('sleep');
    expect(c).toHaveProperty('trainingLoad');
    expect(c).toHaveProperty('weeklyPulse');
    expect(c).toHaveProperty('gutFeeling');
  });
});

File: frontend/mobile/lib/features/recovery/domain/services/recovery_service_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:way2move/features/recovery/domain/entities/recovery_score.dart';
import 'package:way2move/features/recovery/domain/services/recovery_service.dart';
import 'package:way2move/features/sleep/domain/entities/sleep_log.dart';
import 'package:way2move/features/sessions/domain/entities/session.dart';
import 'package:way2move/features/assessments/domain/entities/assessment.dart';
import 'package:way2move/features/nutrition/domain/entities/meal.dart';

void main() {
  final now = DateTime.now();

  // ── Helpers ────────────────────────────────────────────────────────────────

  SleepLog makeSleep({required int quality}) => SleepLog(
        id: 'sl-$quality',
        userId: 'u1',
        bedTime: now.subtract(const Duration(hours: 8)),
        wakeTime: now,
        quality: quality,
        date: now,
      );

  Session makeSession({
    required DateTime date,
    SessionStatus status = SessionStatus.completed,
  }) =>
      Session(
        id: 'sess-${date.millisecondsSinceEpoch}',
        userId: 'u1',
        date: date,
        status: status,
        exerciseBlocks: const [],
      );

  WeeklyPulse makePulse({
    required int energy,
    required int soreness,
    required int motivation,
    required int sleep,
  }) =>
      WeeklyPulse(
        id: 'wp',
        userId: 'u1',
        date: now,
        energyScore: energy,
        sorenessScore: soreness,
        motivationScore: motivation,
        sleepQualityScore: sleep,
      );

  Meal makeMeal({required int stomach}) => Meal(
        id: 'm-$stomach',
        userId: 'u1',
        date: now,
        mealType: MealType.lunch,
        description: 'meal',
        stomachFeeling: stomach,
      );

  // ── Sleep component ────────────────────────────────────────────────────────

  group('RecoveryService._sleepComponent', () {
    test('no sleep data returns neutral 50', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.sleepComponent, 50.0);
    });

    test('quality 1 maps to 0', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [makeSleep(quality: 1)],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.sleepComponent, 0.0);
    });

    test('quality 5 maps to 100', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [makeSleep(quality: 5)],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.sleepComponent, 100.0);
    });

    test('quality 3 maps to 50', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [makeSleep(quality: 3)],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.sleepComponent, 50.0);
    });

    test('averages multiple sleep logs', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [makeSleep(quality: 1), makeSleep(quality: 5)],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.sleepComponent, 50.0);
    });
  });

  // ── Training load component ────────────────────────────────────────────────

  group('RecoveryService._trainingLoadComponent', () {
    test('no sessions returns 75 (well-rested)', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.trainingLoadComponent, 75.0);
    });

    test('no load recently and previously returns 75', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [
          makeSession(
            date: now.subtract(const Duration(days: 10)),
          )
        ],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.trainingLoadComponent, 75.0);
    });

    test('higher recent load than baseline lowers score', () {
      // Heavy last 3 days, nothing in previous 4
      final c1 = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [
          makeSession(date: now.subtract(const Duration(days: 1))),
          makeSession(date: now.subtract(const Duration(days: 2))),
          makeSession(date: now.subtract(const Duration(days: 3))),
        ],
        weeklyPulses: [],
        meals: [],
      );

      // Light last 3 days, more in previous 4
      final c2 = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [
          makeSession(date: now.subtract(const Duration(days: 4))),
          makeSession(date: now.subtract(const Duration(days: 5))),
          makeSession(date: now.subtract(const Duration(days: 6))),
          makeSession(date: now.subtract(const Duration(days: 7))),
        ],
        weeklyPulses: [],
        meals: [],
      );

      expect(c2.trainingLoadComponent, greaterThan(c1.trainingLoadComponent));
    });

    test('planned sessions are not counted', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [
          makeSession(
            date: now.subtract(const Duration(days: 1)),
            status: SessionStatus.planned,
          ),
        ],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.trainingLoadComponent, 75.0);
    });
  });

  // ── Weekly pulse component ─────────────────────────────────────────────────

  group('RecoveryService._weeklyPulseComponent', () {
    test('no pulses returns neutral 50', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.weeklyPulseComponent, 50.0);
    });

    test('all scores at max (5) maps to 100', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [
          makePulse(energy: 5, soreness: 5, motivation: 5, sleep: 5)
        ],
        meals: [],
      );
      expect(c.weeklyPulseComponent, 100.0);
    });

    test('all scores at min (1) maps to 0', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [
          makePulse(energy: 1, soreness: 1, motivation: 1, sleep: 1)
        ],
        meals: [],
      );
      expect(c.weeklyPulseComponent, 0.0);
    });

    test('mid scores map to 50', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [
          makePulse(energy: 3, soreness: 3, motivation: 3, sleep: 3)
        ],
        meals: [],
      );
      expect(c.weeklyPulseComponent, 50.0);
    });
  });

  // ── Gut feeling component ──────────────────────────────────────────────────

  group('RecoveryService._gutFeelingComponent', () {
    test('no meals returns neutral 50', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [],
        meals: [],
      );
      expect(c.gutFeelingComponent, 50.0);
    });

    test('stomach 1 maps to 0', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [],
        meals: [makeMeal(stomach: 1)],
      );
      expect(c.gutFeelingComponent, 0.0);
    });

    test('stomach 5 maps to 100', () {
      final c = RecoveryService.calculateComponents(
        sleepLogs: [],
        recentSessions: [],
        weeklyPulses: [],
        meals: [makeMeal(stomach: 5)],
      );
      expect(c.gutFeelingComponent, 100.0);
    });
  });

  // ── Weighted sum ───────────────────────────────────────────────────────────

  group('RecoveryService.calculateScore', () {
    test('all zeros produces 0', () {
      const components = RecoveryScoreComponents(
        sleepComponent: 0,
        trainingLoadComponent: 0,
        weeklyPulseComponent: 0,
        gutFeelingComponent: 0,
      );
      expect(RecoveryService.calculateScore(components), 0.0);
    });

    test('all 100 produces 100', () {
      const components = RecoveryScoreComponents(
        sleepComponent: 100,
        trainingLoadComponent: 100,
        weeklyPulseComponent: 100,
        gutFeelingComponent: 100,
      );
      expect(RecoveryService.calculateScore(components), 100.0);
    });

    test('weights are correctly applied (30/40/20/10)', () {
      const components = RecoveryScoreComponents(
        sleepComponent: 100, // 30
        trainingLoadComponent: 0,
        weeklyPulseComponent: 0,
        gutFeelingComponent: 0,
      );
      expect(RecoveryService.calculateScore(components), closeTo(30.0, 0.001));
    });

    test('training load is the heaviest weight (40%)', () {
      const components = RecoveryScoreComponents(
        sleepComponent: 0,
        trainingLoadComponent: 100, // 40
        weeklyPulseComponent: 0,
        gutFeelingComponent: 0,
      );
      expect(RecoveryService.calculateScore(components), closeTo(40.0, 0.001));
    });

    test('score is clamped to 0–100', () {
      const components = RecoveryScoreComponents(
        sleepComponent: 200,
        trainingLoadComponent: 200,
        weeklyPulseComponent: 200,
        gutFeelingComponent: 200,
      );
      expect(RecoveryService.calculateScore(components), 100.0);
    });
  });

  // ── recommendationForScore ─────────────────────────────────────────────────

  group('RecoveryService.recommendationForScore', () {
    test('75 returns green recommendation', () {
      final r = RecoveryService.recommendationForScore(75);
      expect(r, contains('well-recovered'));
    });

    test('74 returns yellow recommendation', () {
      final r = RecoveryService.recommendationForScore(74);
      expect(r, contains('Mild fatigue'));
    });

    test('50 returns yellow recommendation', () {
      final r = RecoveryService.recommendationForScore(50);
      expect(r, contains('Mild fatigue'));
    });

    test('49 returns red recommendation', () {
      final r = RecoveryService.recommendationForScore(49);
      expect(r, contains('High fatigue'));
    });

    test('0 returns red recommendation', () {
      final r = RecoveryService.recommendationForScore(0);
      expect(r, contains('High fatigue'));
    });
  });
}
